MakeMsix.targets

```diff
@@ -0,0 +1,78 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT License. See LICENSE in the project root for license information. -->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <!-- Define the property MakeMsixVerbosity=high for additional logging (default=normal) -->
+  <PropertyGroup>
+    <MakeMsixVerbosity Condition="'$(MakeMsixVerbosity)' == ''">normal</MakeMsixVerbosity>
+    <MakeMsixSigned Condition="'$(MakeMsixSigned)' == ''">true</MakeMsixSigned>
+  </PropertyGroup>
+
+  <Target Name="MakeMSIX"
+    Inputs="@(MakeMsixInputs);@(MakeMsixInputsWithLocations);@(MakeMsixInputAssets);$(MakeMsixOutputFilename)"
+    Outputs="$(OutDir)$(TargetName)"
+    AfterTargets="AfterBuild">
+
+    <!-- Initial state diagnostics -->
+    <Message Text="MakeMSIX Pre: MSBuildProjectName=$(MSBuildProjectName)" Importance="$(MakeMsixVerbosity)" />
+    <Message Text="MakeMSIX Pre: ProjectName=$(ProjectName)" Importance="$(MakeMsixVerbosity)" />
+    <Message Text="MakeMSIX Pre: TargetName=$(TargetName)" Importance="$(MakeMsixVerbosity)" />
+    <Message Text="MakeMSIX Pre: MakeMsixInputs=@(MakeMsixInputs)" Importance="$(MakeMsixVerbosity)" />
+    <Message Text="MakeMSIX Pre: MakeMsixInputsWithLocations=@(MakeMsixInputsWithLocations)" Importance="$(MakeMsixVerbosity)" />
+    <Message Text="MakeMSIX Pre: MakeMsixInputAssets=@(MakeMsixInputAssets)" Importance="$(MakeMsixVerbosity)" />
+    <Message Text="MakeMSIX Pre: MakeMsixOutputFilename=$(MakeMsixOutputFilename)" Importance="$(MakeMsixVerbosity)" />
+
+    <!-- Ensure MakeMsixOutputFilename is defined (defining it here if necessary, as per Microsoft.Cpp.Default.props line ~189) -->
+    <PropertyGroup>
+      <MakeMsixOutputFilename Condition="'$(MakeMsixOutputFilename)' == ''">$(ProjectName).msix</MakeMsixOutputFilename>
+    </PropertyGroup>
+    <Message Text="MakeMSIX: TargetName=$(MakeMsixOutputFilename)" Importance="$(MakeMsixVerbosity)" />
+    <Message Text="MakeMSIX: OutDir=$(OutDir)" Importance="$(MakeMsixVerbosity)" />
+
+    <PropertyGroup>
+      <_MakeMsix_WorkDir>$(OutDir)msix</_MakeMsix_WorkDir>
+      <_MakeMsix_OutDir>$(OutDir)</_MakeMsix_OutDir>
```

Comment:
Please use the already defined $(AppxPackageDir) which is set in EnsureOutputLayout.props.  This creates a centralized place for MSIX output in the build that works for the differently managed pipeline and local dev builds.  The property should already be set, and if it is not, then there is an ordering issue to be diagnosed.

Reply:
I see MrtCore.PriGen.targets does a whole bunch of local definition., but is that generic or specific to MRTCore? If generic shouldn't that be in a .props file imported as needed?

EnsureOutputLayout.props defines AppxPackageDir = $(BaseOutputPath)AppxPackages e.g. /BuildOutput/Debug/x64/AppxPackages. That's equivalent to what the various test projects were doing (MakeMsix.targets just centralizes that, didn't invent the path).

Q: Why plural AppxPackages?

Q: EnsureOutputLayout.props defines AppxPackageDir=...AppxPackages -- singular=plural. Seems odd. Intentional or accidental?

Q: This is the 'workdir' where we collect files to then run makeappx.exe "make a package from that dir and all its children". Should that be an 'Out' or 'Int' directory? IOW should that be under BuildOutput or obj? Seems like the latter. Isn't this just a temporary interim working directory to create the .msix? That once we have the .msix we don't need this workdir anymore? If so then obj would be the better place.

Reply:
@aeloros Here's a weird one. In makemsix.targets add some printfery to the Clean target

```
  <Target Name="CleanMakeMSIX" AfterTargets="AfterClean">
<Message Text="aaa $(OutputPath)" Importance="high" />
<Message Text="bbb $(OutDir)" Importance="high" />
<Message Text="xxx $(MakeMsixOutputFilename)" Importance="high" />
<Message Text="yyy $(TargetName)" Importance="high" />
```

Then in VS do Clean Solution and you get inconsistent output:

```
27>------ Clean started: Project: Framework.Math.Add, Configuration: Debug x64 ------
27>aaa C:\source\repos\windowsappsdk\BuildOutput\Debug\x64\Framework.Math.Add\
27>bbb C:\source\repos\windowsappsdk\BuildOutput\Debug\x64\Framework.Math.Add\
27>xxx Framework.Math.Add.msix
27>yyy Framework.Math.Add
```

vs

```
36>------ Clean started: Project: Microsoft.WindowsAppRuntime.Framework, Configuration: Debug x64 ------
36>aaa C:\source\repos\windowsappsdk\BuildOutput\Debug\x64\
36>bbb C:\source\repos\windowsappsdk\BuildOutput\Debug\x64\Microsoft.WindowsAppRuntime.Framework\
36>xxx Microsoft.WindowsAppRuntime.Framework.msix
36>yyy Microsoft.WindowsAppRuntime.Framework
```

Observations:

* OutputPath=OutDir in 3 projects (Framework.Math.Add, Framework.Math.Multiply, Framework.Widgets)
* OutputPath!=OutDir in the other 10 projects
* OutputPath is the root-of-all-output vs OutDir is a project-specific output
* Per `build\EnsureOutputLayout.props` I'd expect the exact opposite

```
    <BaseOutputPath>$(RepoRoot)\BuildOutput\$(Configuration)\$(PlatformTarget)\</BaseOutputPath>
...
    <OutDir>$(BaseOutputPath)</OutDir>
    <OutputPath>$(OutDir)\$(MSBuildProjectName)\</OutputPath>
```

OutputPath should point to a subdir of OutDir, but I'm seeing the exact opposite. I don't understand why. Do you?